### ISSUE TITLE
isinstance(x, NoneType) doesn't get checked if the type annotation is None

### ISSUE THREAD

**Issue by strny0** [python/mypy#13154]
**Bug Report**

isinstance check of mypy doesn't detect instances of NoneType (actual None), doesn't visit/check branches checked of  `isinstance()` conditions.

**Expected Behavior**

```python
from types import NoneType

def f(x: str | None):
    if isinstance(x, NoneType):
        reveal_type(x) # N: Revealed type is "None"
    elif isinstance(x, str):
        reveal_type(x) # N: Revealed type is "builtins.str"
    reveal_type(x) # N: Revealed type is "Union[builtins.str, None]"
```

**Actual Behavior**

```python
from types import NoneType

def f(x: str | None):
    if isinstance(x, NoneType):
        reveal_type(x)
    elif isinstance(x, str):
        reveal_type(x) # N: Revealed type is "builtins.str"
    reveal_type(x) # N: Revealed type is "Union[builtins.str, None]"
```

**Your Environment**

- Mypy version used: Any version that implements PR #13153 
- Mypy command-line flags: None
- Mypy configuration options from `mypy.ini` (and other config files): None
- Python version used: 3.10
- Operating system and version: Linux 5.15.0-41-generic 20.04.1-Ubuntu x86_64

**Comment by erictraut** [comment]
This is a pretty atypical way to check for `None`. The more idiomatic (and easier-to-read) takes advantage of the fact that `None` is a singleton.

```python
def f(x: str | None):
    if x is None:
        ...
```

**Comment by KotlinIsland** [comment]
basedmypy handles this correctly:
```py
from types import NoneType

def f(x: str | None):
    if isinstance(x, NoneType):
        reveal_type(x) # N: Revealed type is "None"
    elif isinstance(x, str):
        reveal_type(x) # N: Revealed type is "str"
    reveal_type(x) # N: Revealed type is "str | None"
```

**Comment by KotlinIsland** [comment]
This would fix it: https://github.com/python/mypy/issues/11287

**Comment by tyralla** [comment]
@strny0 We implemented this in Mypy 1.9, but I wasn't aware of this issue.  I think it can be closed.

**Comment by JelleZijlstra** [comment]
Confirmed mypy's behavior is now the expected behavior in the OP.
